pipenv/core.py

```diff
@@ -351,6 +351,42 @@ def find_a_system_python(line):
     return python_entry
 
 
+def find_python_installer(name, env_var):
+    """
+    Given a python installer (pyenv or asdf), try to locate the binary for that
+    installer.
+
+    pyenv/asdf are not always present on PATH. Both installers also support a
+    custom environment variable (PYENV_ROOT or ASDF_DIR) which alows them to
+    be installed into a non-default location (the default/suggested source
+    install location is in ~/.pyenv or ~/.asdf).
+
+    For systems without the installers on PATH, and with a custom location
+    (e.g. /opt/pyenv), Pipenv can use those installers without modifications to
+    PATH, and only with their respective environment variables in a .env file.
+    This is desirable, since setting PATH in an .env file is annoying due to
+    the common need to reference the pre-existing PATH variable, which is not
+    supported in .env.
+
+    This function searches for installer binaries using PATH, their respective
+    environment variables, and their respective default install locations. This
+    allows Pipenv to use those installers regardless of shell configuration, so
+    long as PYENV_ROOT or ASDF_DIR is specified in an .env file.
+    """
+    for candidate in (
+        # Look for the Python installer using the equivalent of 'which'. On
+        # Homebrew-installed systems, the env var may not be set, but this
+        # strategy will work.
+        find_windows_executable('', name),
+        # Check for explicitly set install locations (e.g. PYENV_ROOT, ASDF_DIR).
+        os.path.join(os.path.expanduser(os.getenv(env_var, '/dev/null')), 'bin', name),
+        # Check the pyenv/asdf-recommended from-source install locations
+        os.path.join(os.path.expanduser('~/.{}'.format(name)), 'bin', name),
```

Comment:
surfacing the installers in explicit variables is desirable over burying them in a function because it becomes easier to reason about what is happening, or to separate out the location of the install as compared with the process for finding python executables

Reply:
I can refactor the code to make the search paths more explicit, would that work?

Reply:
In general, `pythonfinder` is insufficient for this task. It surfaces code that checks whether or not `pyenv`/`asdf` are *in use* (on the invoking shell), but it provides no facilities for figuring out where the `pyenv`/`asdf` binaries *actually are*.

`pythonfinder` does a good job of its namesake (finding Python), but does not contain any facilities for finding python installers.

I suppose I could submit a PR to `pythonfinder`, but that would be a brand-new capability for that package, and I suspect a bit out of scope of its goals. Pipenv, which needs to *invoke* `pyenv`/`asdf` is a better place for that logic to live, I think.

Reply:
I definitely don't want to add the installer capability to pythonfinder, i think it's fine to add here. Overall I don't see any major issues with this as-is and think it's fine to leave